File: problems/Python/e004.py
''' Find the largest palindrome made from the product of two 3-digit numbers '''

def main():
    ''' Generate all possible products of 3-digit numbers, keep track of
    largest palindrome '''
    cur_max = 0
    for n1 in range(999, 99, -1):
        for n2 in range(999, 99, -1):
            p = n1*n2
            if p < cur_max:
                continue
            if str(p) == str(p)[::-1]:
                cur_max = p
    print(cur_max)


if __name__ == "__main__":
    main()

File: problems/C/e004.c
// Find the largest palindrome made from the product of two 3-digit numbers

#include <stdio.h>
#include <stdbool.h>
#include <math.h>

bool isPalindrome(int num);

int main(void)
{
    int curMax = 0;
    for (int i = 999; i > 99; i--)
        for (int j = i; j > 99; j--)
        {
            int p = i*j;
            if (p > curMax && isPalindrome(p))
                curMax = p;
        }

    printf("%i\n", curMax);
    return 0;
}

bool isPalindrome(int num)
{
    int left = pow(10, (int)log10(num));

    while (num)
    {
        if (num / left != num % 10)
            return false;
        num %= left;
        num /= 10;
        left /= 100;
    }

    return true;
}
